### ISSUE TITLE
2.9.8 started serializing my uuid's as arrays instead of strings

### ISSUE THREAD

**Issue by stijnfrishert** [async-graphql/async-graphql#603]
I have a custom newtype that contains a [`Uuid`](https://docs.rs/uuid/0.8.2/uuid/), that basically looks like this:

```
#[derive(Serialize, Deserialize)]
#[repr(transparent)]
#[serde(transparent)]
AssetId(pub uuid::Uuid)
```

This is then wrapped into an `async_graphql::scalar!(AssetId)`.

- In versions below `2.9.8` this used to serialize (to JSON) as a hyphenated string, such as `"3292da00-ff2c-45aa-9d8e-741f6a3b0b0c"`.
- Starting from `2.9.8`, this is serialized as a byte array, such as `[113, 225, ...]`.
- Looks like the `Serializer` used in `2.9.8` isn't human readable, which leads `Uuid` to use the byte array serialization.

What changed? And more importantly, why did that change? `CHANGELOG.md` doesn't mention anything about Uuid's or serialization, and this seems like a pretty big breaking change (warranting a major version upgrade).

**Comment by Miaxos** [comment]
You sure it's the `2.9.8` which is not working for your use case?

I tried to create a reproductible use case, and from my findings, it's this commit from the https://github.com/async-graphql/async-graphql/commit/ec8f97ba9e1c37433742dc5b8f9df1a02dfb9ec9  (2.9.9 release) which is failing with the implementation of `Deserialize` and `Serialize` instead of the trait auto-generation https://github.com/async-graphql/async-graphql/commit/ec8f97ba9e1c37433742dc5b8f9df1a02dfb9ec9#diff-592c2b8e5e98ef9b09e52148d90a976f0153e3f0659a0489725f0ce26669da3aL124-L125

The test case https://github.com/async-graphql/async-graphql/pull/604

**Comment by stijnfrishert** [comment]
Double checked for you @Miaxos, but looks like you're right. I switch between versions `=2.9.7` and `=2.9.8`, assuming I'd get exactly [the specified version](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility), but looking at the `Cargo.lock` it jumped up to `2.9.10`.

Thanks for taking a look so quickly. Sounds like CBOR won't be a problem as it shouldn't be `human_readable`, but I'll leave that to the pro's ;)

**Comment by Erik1000** [comment]
> Are you sure it's the `2.9.8` which is not working for your use case?
> 
> 
> 
> I tried to create a reproductible use case, and from my findings, it's this commit from the https://github.com/async-graphql/async-graphql/commit/ec8f97ba9e1c37433742dc5b8f9df1a02dfb9ec9  (2.9.9 release) which is failing with the implementation of `Deserialize` and `Serialize` instead of the trait auto-generation https://github.com/async-graphql/async-graphql/commit/ec8f97ba9e1c37433742dc5b8f9df1a02dfb9ec9#diff-592c2b8e5e98ef9b09e52148d90a976f0153e3f0659a0489725f0ce26669da3aL124-L125
> 
> 
> 
> The test case https://github.com/async-graphql/async-graphql/pull/604
> 
> 
> 
> A possible solution would be to implement the Scalar trait without using `scalar!`.
> 
> Well, I don't think it was an intended breaking changes, it's more a hidden and annoying bug.
> 
> 
> 
> **EDIT**: I proposed a fix too, with the `human_readable` to prefer construct a string instead of bytes if possible, but idk, it may be an issue for those using CBOR. (👋  @sunli829 )

@Miaxos with serde_cbor uuid::Uuid actually serialises to bytes instead of a string. However, the UUID scalar behind the uuid feature always serialises to a String @stijnfrishert 😉

**Comment by Miaxos** [comment]
> However, the UUID scalar behind the uuid feature always serialises to a String

Not at that time, there was a breaking changes with a manual implementation of a Serializer for the human_readable and not a generated by macro one which serialized uuid to Bytes instead of string, the serialization of an uuid can be both String / Bytes, it'll depends on the is_human_readable function from the serializer implementation https://docs.serde.rs/serde/trait.Serializer.html#method.is_human_readable

**Comment by sunli829** [comment]
I will close this issue, if you have other questions, please feel free to open it.

**Comment by anlumo** [comment]
There's an inconsistency there now. `Serializer` returns `true` for `is_human_readable` due to #604, but Deserializer is still returning `false`:

https://github.com/async-graphql/async-graphql/blob/8dc6fec0fc65605bb5d67ee8bfc3aa85339fcd3c/value/src/deserializer.rs#L197-L199

This means that sending Uuids server to client works fine, but in the other direction the server expects a binary value (so it can be deserialized), which I don't seem to be able to send via JSON.

**Comment by Miaxos** [comment]
Good point! I'll look at this tomorrow

**Comment by stijnfrishert** [comment]
Yes, just tried with `2.11.3`, but the issue still seems doesn't seem to be fixed :)

**Comment by stijnfrishert** [comment]
Yup, tried this with `2.11.3` today and sending ids to the server still seems broken

**Comment by sunli829** [comment]
I fixed it in `3.0`.🙂
